components/automate-cli/pkg/verifyserver/services/fqdnservice/fqdnservice.go

```diff
@@ -0,0 +1,240 @@
+	"github.com/chef/automate/lib/logger"
+)
+
+type IFqdnService interface {
+	CheckFqdnReachability(models.FqdnRequest, string) models.FqdnResponse
+}
+
+type FqdnService struct {
+	log     logger.Logger
+	timeout time.Duration
+}
+
+func NewFqdnService(log logger.Logger, timeout time.Duration) IFqdnService {
+	return &FqdnService{
+		log:     log,
+		timeout: timeout,
+	}
+}
+
+func (fq *FqdnService) createClient(rootCert string) *http.Client {
+	fq.log.Debug("Creating Client")
+	caCertPool := x509.NewCertPool()
+	caCertPool.AppendCertsFromPEM([]byte(rootCert))
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs:    caCertPool,
+				MinVersion: tls.VersionTLS12,
+			},
+		},
+		Timeout: fq.timeout * time.Second,
+	}
+	fq.log.Debug("Client Created.")
+	return client
+}
+
+func createErrorMessage(setNodes map[string]int, reqnodes []string, isAfterDeployment bool) string {
+	temp := "["
+	for _, k := range reqnodes {
+		var val string
+		if isAfterDeployment {
+			hasher := md5.New() // nosemgrep
+			_, err := io.WriteString(hasher, k)
+			if err != nil {
+				return constants.IP_TO_HASH_FAIL_MESSAGE
+			}
+			val = hex.EncodeToString(hasher.Sum(nil))
+		} else {
+			val = k
+		}
+
+		if setNodes[val] > 0 {
+			temp += k
+			temp += ", "
+		}
+	}
+	temp = strings.TrimSuffix(temp, ", ")
+	temp += "]"
+
+	return fmt.Sprintf(constants.NODE_ERROR_MESSAGE, temp)
+}
+
+// makeSet function is storing all the request nodes into the map according to the deployment.
+// if isAfterDeployment is true then it will store the hash of the ips of request nodes,
+// otherwise it will directly store the ips.
+func makeSet(reqNodes []string, isAfterDeployment bool) (map[string]int, error) {
+	setNodes := make(map[string]int)
+	for _, k := range reqNodes {
+		if isAfterDeployment {
+			hasher := md5.New() // nosemgrep
+			_, err := io.WriteString(hasher, k)
+			if err != nil {
+				return nil, err
+			}
+			setNodes[hex.EncodeToString(hasher.Sum(nil))] += 1
+		} else {
+			setNodes[k] += 1
+		}
+	}
+
+	return setNodes, nil
+}
+
+// fqdnReachable function will check that are we able to hit the load balancer fqdn or not.
+func (fq *FqdnService) fqdnReachable(fqdn, rootCert, nodeType string, isAfterDeployment bool, port string) models.Checks {
+	fq.log.Debug("Checking Fqdn Reachability...")
+	client := fq.createClient(rootCert)
+	var url string
+
+	if isAfterDeployment && nodeType == constants.CHEF_INFRA_SERVER {
+		url = fmt.Sprintf("https://%s:%s/_status", fqdn, port)
+	} else {
+		url = fmt.Sprintf("https://%s:%s", fqdn, port)
+	}
+	fq.log.Debug("URL: ", url)
+
+	res, err := client.Get(url)
+	if err != nil {
+		fq.log.Error(err.Error())
+		return createCheck(constants.FQDN_TITLE, false, "", constants.FQDN_ERROR_MESSAGE, constants.FQDN_RESOLUTION_MESSAGE)
+	}
+
+	fq.log.Debug("Status Code: ", res.StatusCode)
+	if res.StatusCode != 200 {
+		fq.log.Debug("Fqdn is not reachable.")
```

Comment:
We can have fqdn here also, that which fqdn is not reachable.

Reply:
done